---
title: "BCB546 R Assignment"
author: "Brittany Cavazos"
date: "September 30, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part One

## Data Inspection
First I need to read in the data and briefly look over the structure of the datasets
```{r reading in data}
genotypes_raw<-read.delim("fang_et_al_genotypes.txt", header = T, sep = "\t")
dim(genotypes_raw)
head(genotypes_raw[,1:5]) # just showing the output of first five columns for all of these inspections because there are a ton of columns..
tail(genotypes_raw[,1:5])
# str(genotypes_raw) output is pretty long
snp_pos<-read.delim("snp_position.txt", header = T, sep = "\t")
dim(snp_pos)
head(snp_pos[,1:5])
tail(snp_pos[,1:5])
# str(snp_pos)
```

Next I need to transpose the data and subset to maize and teosinte datasets. 
For maize: Group = ZMMIL, ZMMLR, and ZMMMR 
```{r}
summary(genotypes_raw$Group)
  # for ZMMIL, ZMMLR and ZMMMR, we want a total maize subset that is 1573 rows long
maize_raw<-dplyr::filter(genotypes_raw, Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR")
  # there is probaby a shorter way of doing this but it works

  # now to transpose it to have SNP ID as columns
names(maize_raw)[1:25]
maize_T<-as.data.frame(t(maize_raw))
dim(maize_T)
Maize_IDs<-as.character(maize_raw$Sample_ID)
colnames(maize_T)<-Maize_IDs
```

For teosinte: Group = ZMPBA, ZMPIL, and ZMPJA 
```{r}
teosinte_raw<-dplyr::filter(genotypes_raw, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
  # there is probaby a shorter way of doing this but it works
  # should be 975 rows after looking at summary(genotypes$Group)
dim(teosinte_raw)
  # not to transpose it to have SNP ID as columns
names(teosinte_raw)[1:25]
teosinte_T<-as.data.frame(t(teosinte_raw))
dim(teosinte_T)
Teosinte_IDs<-as.character(teosinte_raw$Sample_ID)
colnames(teosinte_T)<-Teosinte_IDs # after transposing we got weird column names so this is a way to fix that

```
Now I need to join together the snp_pos with the other two data sets, but I only need a couple columns from the snp data set
```{r}
names(snp_pos)
snp_pos<-dplyr::select(snp_pos, SNP_ID, Chromosome, Position)
  # usally writing over data frames isn't great but this is probably fine

```
Now I need to prep the data sets for merges - we will join maize/teosinte with snp_pos by snp id
```{r}

# need to add a column in both maize and teosinte that is the snp id - currently its only as the rowname
row.names(snp_pos)<-as.character(snp_pos$SNP_ID)

maize_T_forjoin<-maize_T[4:986,] # to join the snpid info we have to get rid of some of the sample info, we can tack this back on at the end if we need to
teosinte_T_forjoin<-teosinte_T[4:986,]

maizejoined<-merge(snp_pos, maize_T_forjoin, by = "row.names", all = F)
dim(maizejoined)
maizejoinedF<-dplyr::select(maizejoined, -Row.names)
teosintejoined<-merge(snp_pos, teosinte_T_forjoin, by = "row.names", all = F)
dim(teosintejoined)
teosintejoinedF<-dplyr::select(teosintejoined, -Row.names) # F for final

```

## Data Processing
_For Maize_
Now we need to make 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ? 
# 
```{r}
head(maizejoinedF[maizejoinedF$Chromosome == 1,]) # they are currently out of order

# first let's make chromosome and position numeric
maizejoinedF$Chromosome<-as.numeric(maizejoinedF$Chromosome)
maizejoinedF$Position<-as.numeric(maizejoinedF$Position)
# sort by chromosome, then position
library(dplyr) # not totally sure if this is needed or not
maizejoined_sorted <- maizejoinedF %>% group_by(Chromosome)
maizejoined_sorted <- maizejoined_sorted %>% arrange(Position, .by_group = T)
```

Making the for loop by subsetting the data by already ordered chromosomes
```{r}
#this creates files that are pretty big and actually ended up crashing rstudio... they are commented out for now
#for (i in 1:10) {
#  Chr_i<- dplyr::filter(maizejoined_sorted, Chromosome == i)
#  filename <- paste("MaizeChromosome", i, ".txt", sep="")
#  write.table(Chr_i, file = filename, sep = "\t")
#}
```


To make the second 10 files in the reverse order with missing data encoded by "-", we need to reverse the sorting and replace ? with -

```{r}
maizejoined_reverse <- maizejoinedF %>% group_by(Chromosome)
maizejoined_reverse <- maizejoined_reverse %>% arrange(desc(Position), .by_group = T)
head(maizejoined_reverse)

# make all of the SNP columns into characters 
#as.character(maizejoined_reverse[,4])

maizejoined_reverse[,4:1576]<-(as.character(unlist(maizejoined_reverse[,4:1576])))
tail(maizejoined_reverse[,4:1576], n=2)
maize_reversefinal<-as.data.frame(lapply(maizejoined_reverse, function(x) {gsub("\\?","-", x)}))

tail(test[,4:1576], n=2)

```
Now to write these out 
```{r}
#for (i in 1:10) {
#  Chr_i<- dplyr::filter(maize_reversefinal, Chromosome == i)
#  filename <- paste("MaizeChr_reverseorder", i, ".txt", sep="")
#  write.table(Chr_i, file = filename, sep = "\t")
#}
```

_For Teosinte_
For teosinte, we will follow the same steps as in the maize files
Now we need to make 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ? 
# 
```{r}
head(teosintejoinedF[teosintejoinedF$Chromosome == 1,]) # they are currently out of order

# first let's make chromosome and position numeric
teosintejoinedF$Chromosome<-as.numeric(teosintejoinedF$Chromosome)
teosintejoinedF$Position<-as.numeric(teosintejoinedF$Position)
# sort by chromosome, then position
library(dplyr) # not totally sure if this is needed or not
teosintejoined_sorted <- teosintejoinedF %>% group_by(Chromosome)
teosintejoined_sorted <- teosintejoined_sorted %>% arrange(Position, .by_group = T)
```

Making the for loop by subsetting the data by already ordered chromosomes
```{r}
#this creates files that are pretty big and actually ended up crashing rstudio... they are commented out for now
#for (i in 1:10) {
#  Chr_i<- dplyr::filter(teosintejoined_sorted, Chromosome == i)
#  filename <- paste("TeosinteChromosome", i, ".txt", sep="")
#  write.table(Chr_i, file = filename, sep = "\t")
#}
```


To make the second 10 files in the reverse order with missing data encoded by "-", we need to reverse the sorting and replace ? with -

```{r}
teosintejoined_reverse <- teosintejoinedF %>% group_by(Chromosome)
teosintejoined_reverse <- teosintejoined_reverse %>% arrange(desc(Position), .by_group = T)
head(teosintejoined_reverse)

# make all of the SNP columns into characters 
#as.character(maizejoined_reverse[,4])

teosintejoined_reverse[,4:978]<-(as.character(unlist(teosintejoined_reverse[,4:978])))
tail(teosintejoined_reverse[,4:978], n=2)
teosinte_reversefinal<-as.data.frame(lapply(teosintejoined_reverse, function(x) {gsub("\\?","-", x)}))

tail(teosinte_reversefinal[,4:1576], n=2)

```
Now to write these out 
```{r}
#for (i in 1:10) {
#  Chr_i<- dplyr::filter(teosinte_reversefinal, Chromosome == i)
#  filename <- paste("TeosinteChr_reverseorder", i, ".txt", sep="")
#  write.table(Chr_i, file = filename, sep = "\t")
#}
```

# Part Two
Use ggplot to visualize our data in this part

First, I need to make tidy datasets for both corn and teosinte
```{r}
library(reshape2); library(ggplot2)
head(snp_pos)
# I wrote over the original snp_pos to make it only 3 variables so here I'm going back to the orig
SNPs<-read.delim("snp_position.txt", header = T, sep = "\t")

snpMelt<-melt(SNPs, "Chromosome", "SNP_ID")
head(snpMelt)
# make chrom. integer so it plots in order
SNPs$Chromosome<-factor(SNPs$Chromosome, levels = c(1:10, "multiple", "unknown"))

ggplot(SNPs) + geom_bar(aes(x=Chromosome)) + theme_bw() + labs(x="Chromosome",y="SNP Count") 

# SNPS per chromosome 
# Plot the total number of SNPs in our dataset on each chromosome. What groups contribute most of these SNPs? 
View(head(genotypes_raw))
```
Chromosome 1 seems to have the most snps 

```{r}
# Tidy data
tidy_genotypes<-melt(genotypes_raw, c(1,3), c(4:986), variable.name = "SNP_ID", value.name = "SNP") #melting snps by sample id and keeping group
head(tidy_genotypes, n = 40)


SiteStatus <- function(x){
  if(x == "?/?")
    return("NA")
  else if ( x == "A/A" | x == "T/T"| x == "C/C" | x == "G/G") 
    return("homozygous")
  else
    return("heterozygous")
}

tidy_genotypes$sitestatus<-as.character(lapply(tidy_genotypes$SNP, FUN = SiteStatus))

# it worked!


```
